File: src/main/java/me/thesourcecode/sourcebot/commands/administrator/ReactCommand.java
package me.thesourcecode.sourcebot.commands.administrator;

import me.thesourcecode.sourcebot.api.Source;
import me.thesourcecode.sourcebot.api.command.Command;
import me.thesourcecode.sourcebot.api.command.CommandInfo;
import me.thesourcecode.sourcebot.api.entity.SourceRole;
import me.thesourcecode.sourcebot.api.message.alerts.CriticalAlert;
import me.thesourcecode.sourcebot.api.message.alerts.SuccessAlert;
import net.dv8tion.jda.api.JDA;
import net.dv8tion.jda.api.MessageBuilder;
import net.dv8tion.jda.api.entities.*;
import net.dv8tion.jda.api.exceptions.ErrorResponseException;

public class ReactCommand extends Command {
    private final CommandInfo INFO = new CommandInfo(
            "react",
            "Makes source add a reaction to all current reactions on a message.",
            "(channel id) <message id>",
            CommandInfo.Category.ADMIN)
            .asGuildOnly()
            .withControlRoles(SourceRole.ADMIN, SourceRole.OWNER);

    @Override
    public CommandInfo getInfo() {
        return INFO;
    }

    @Override
    public Message execute(Source source, Message message, String[] args) {
        final JDA jda = source.getJda();
        final User user = message.getAuthor();
        final MessageChannel messageChannel = message.getChannel();

        final String channelId = args.length == 1 ? messageChannel.getId() : args[0];
        final String messageId = args.length == 1 ? args[0] : args[1];

        final TextChannel reactChannel = jda.getTextChannelById(channelId);
        if (reactChannel == null) {
            CriticalAlert alert = new CriticalAlert();
            alert.setTitle("Uh Oh!").setDescription("I could not find a text channel with the given channel id!");
            return new MessageBuilder(alert.build(user)).build();
        }

        final Message reactMessage;

        try {
            reactMessage = reactChannel.retrieveMessageById(messageId).complete();
        } catch (ErrorResponseException ex) {
            CriticalAlert alert = new CriticalAlert();
            alert.setTitle("Uh Oh!")
                    .setDescription("I could not find a message in the given text channel with that id!");
            return new MessageBuilder(alert.build(user)).build();
        }

        reactMessage.getReactions().forEach(reaction -> {
            final MessageReaction.ReactionEmote reactionEmote = reaction.getReactionEmote();
            try {
                reactMessage.addReaction(reactionEmote.getEmote()).queue();
            } catch (IllegalStateException ex) {
                reactMessage.addReaction(reactionEmote.getName()).queue();
            }
        });

        SuccessAlert alert = new SuccessAlert();
        alert.setDescription("Successfully reacted to all reactions on the given message!");
        return new MessageBuilder(alert.build(user)).build();
    }
}

File: API/src/main/kotlin/net/sourcebot/impl/command/ReactCommand.kt
package net.sourcebot.impl.command

import me.hwiggy.kommander.arguments.Adapter
import me.hwiggy.kommander.arguments.Arguments
import me.hwiggy.kommander.arguments.Synopsis
import net.dv8tion.jda.api.entities.Message
import net.dv8tion.jda.api.entities.TextChannel
import net.sourcebot.api.command.RootCommand
import net.sourcebot.api.command.argument.SourceAdapter
import net.sourcebot.api.response.Response
import net.sourcebot.api.response.StandardSuccessResponse

class ReactCommand : RootCommand() {
    override val name = "react"
    override val description = "React to a message's reactions as the bot."
    override val permission = name
    override val guildOnly = true

    override val synopsis = Synopsis {
        optParam("channel", "The channel the message is in.", Adapter.single())
        reqParam("message", "The message ID of the message.", Adapter.single())
    }

    override fun execute(sender: Message, arguments: Arguments.Processed): Response {
        val channel = arguments.optional<String, TextChannel>("channel", sender.textChannel) {
            SourceAdapter.textChannel(sender.guild, it)
        }
        val messageId = arguments.required<String>("message", "You did not specify a message ID to react to!")
        val retrieved = channel.retrieveMessageById(messageId).complete()
        retrieved.reactions.forEach {
            val reaction = it.reactionEmote
            when {
                reaction.isEmoji -> retrieved.addReaction(reaction.emoji)
                reaction.isEmote -> retrieved.addReaction(reaction.emote)
                else -> (throw IllegalStateException("Invalid emote type discovered!"))
            }.queue()
        }
        return StandardSuccessResponse(
            description = "Successfully reacted to all reactions for the given message!"
        )
    }
}